File: solutions/esab-atad/rust/ESAb_ATAd.rs
use std::process;
use std::io::{self, Write};

fn main(){
    let mut input = String::new();
    io::stdout().flush();
     //println!("Say something");
    match io::stdin().read_line(&mut input){
        Ok(_) => {
          //println!("Succes, you said {}", input);
            
      },
      Err(e) => process::exit(1)
    }
    let mut split = input.split_whitespace();
    let vec = split.collect::<Vec<&str>>();
    let mut T = vec[0].parse::<i32>().unwrap();
    let mut B = vec[1].parse::<i32>().unwrap();
  
    for i in 0..T {
        esab_atad(B as usize);   
    }
}

fn esab_atad(b: usize){
    let mut result =  Vec::<i32>::with_capacity(b+1);
    for i in 0..(b+1) {
    result.push(0);
  } 
    let mut L = 1;
    let mut R = b;
    let mut nr = 1;
    while true {
        //println!("nr del for {}",nr);
        //println!("L {}",L);
        //println!("R {}",R);
        if nr%10 == 1 && nr != 1{
            //println!("cambia algo");
            let mut found: i32 = -1;
            let mut found_diff = -1;
            for i in 1..L {
                //println!("i en primer for {}", i);
                if result[i as usize] == result[(b+1-i) as usize]{
                    found = i as i32;
                }else{
                    found_diff = i as i32;
                }
            }
            if(found == -1){
                let mut new_value = query(1);
                query(1);
                if new_value != result[1] {
                    for i in 1..L+1{
                        //println!("i en segundo for {}", i);
                        result[i as usize] ^= 1;
                    }
                    for i in R..b+1{
                       // println!("i en tercer for {}", i);
                        result[i as usize] ^= 1;
                    }
                }
            }else{
                let mut one = query(found);
                if one != result[found as usize]{
                    for i in 1..L+1{
                        //println!("i en cuarto for {}", i);
                        result[i as usize] ^= 1;
                    }
                     for i in R..b+1{
                         //println!("i en quinto for {}", i);
                        result[i as usize] ^= 1;
                    }
                }
                if found_diff == -1{
                    query(found);
                }else{
                    if query(found_diff) != result[found_diff as usize] {
                        result.remove(0);    
                        result.reverse();      
                        result.push(0);    
                        for i in (1..result.len()).rev(){
                            let mut aux = result[(i-1)];
                            result[(i-1)] = result[i];
                            result[i] = aux;
                        } 
                    }
                }
            }
            nr = nr+2;
        }
        result[L as usize] = query(L as i32);
        result[R as usize] = query(R as i32);
        L = L+1;
        R = R-1;
        if L > R {
            let mut output = result.clone();
            let mut output2: String = output.into_iter().map(|i| i.to_string()).collect::<String>();
            output2.remove(0);
            println!("{}",output2);
            io::stdout().flush();
            let mut line2 = String::new();            
            match io::stdin().read_line(&mut line2){
                Ok(_) => {
                    if line2 != "Y" {
                        //process::exit(1);
                        return;
                    }       
                },
                Err(e) => process::exit(1)
            }  
        }
        nr = nr+2;
    }
}

fn query(i: i32) -> i32{
    
    let mut line = String::new();  
    let mut input = 0;
    println!("{}",i);//aqui
    io::stdout().flush();
    match io::stdin().read_line(&mut line){
      Ok(_) => {
                          
      },
      Err(e) =>process::exit(1)
    }  
    io::stdout().flush();
    let mut split = line.split_whitespace();  
    let vec = split.collect::<Vec<&str>>();
    input = vec[0].parse::<i32>().unwrap();
   // println!("return del query {}", input);
    return input;

}

File: solutions/esab-atad/ruby/esab-atad.rb
def query(i)
	#we output the index we want to know, then wait for the judge's response
	puts i
	STDOUT.flush
	r = gets.to_i
	STDIN.flush
	return r
end

def reverseFrom(arr)
	#reverse the given array (arr) starting at index 1
    j = 1
    k = arr.length()-1
    while j < k do
        aux = arr[j]
        arr[j] = arr[k]
        arr[k] = aux
        j+=1
        k-=1
    end
    return arr
end

def test_case()
	answer = Array.new($B+1)
	l = 1
	r = $B
	nr = 1
	while true do
		if nr % 10 == 1 && nr != 1
			found = -1
			found_diff = -1
			for i in 1..l-1 do
				if answer[i] == answer[$B+1-i]
					found = i
				else
					found_diff = i
				end
			end
			if found == -1
				new_value = query(1)
				query(1)
				if new_value != answer[1]
					for i in 1..l do
						answer[i] ^= 1
					end
					for i in r..$B do
						answer[i] ^= 1
					end
				end
			else
				one = query(found)
				if one != answer[found]
					for i in 1..l do
						answer[i] ^= 1
					end
					for i in r..$B do
						answer[i] ^= 1
					end
				end
				if found_diff == -1
					query(found);
				else
					if query(found_diff) != answer[found_diff]
						answer = reverseFrom(answer)
					end
				end
			end
			nr += 2;
		end
		answer[l] = query(l)
		answer[r] = query(r)
		l += 1
		r -= 1
		if l > r
			for i in 1..$B do
				print answer[i]
			end
			print "\n"
			STDOUT.flush
			judge = gets
			STDIN.flush
			if judge.length
    			if judge[0] != 'Y'
    			    exit(0)
    			end
    		else 
    		    exit(0)
    		end
			return
		end
	nr += 2
	end
end

general = gets.split()
STDIN.flush
t = general[0].to_i
$B = general[1].to_i
for rey_sux in 1..t
    test_case()
end